crates/turbo-tasks-memory/src/cell.rs

```diff
@@ -1,37 +1,291 @@
+            Cell::Full(box FullCell::Recomputing { event, .. }) => {
+                let listener = event.listen_with_note(note);
+                Err(RecomputingCell {
+                    listener,
+                    schedule: false,
+                })
+            }
+            Cell::InitialValue { content, .. }
+            | Cell::Full(box FullCell::UpdatedValue { content, .. }) => Ok(content.clone()),
+        }
+    }
+
+    /// INVALIDATION: Be careful with this, it will not track dependencies, so
+    /// using it could break cache invalidation.
+    pub fn read_own_content_untracked(&self) -> CellContent {
+        match self {
+            Cell::Empty
+            | Cell::Recomputing { .. }
+            | Cell::Full(box FullCell::Recomputing { .. }) => CellContent(None),
+            Cell::InitialValue { content, .. }
+            | Cell::Full(box FullCell::UpdatedValue { content, .. }) => content.clone(),
+        }
     }
 
     pub fn track_read(&mut self, reader: TaskId) {
-        self.dependent_tasks.insert(reader);
+        match self {
+            Cell::Empty => {}
+            &mut Cell::Recomputing {
+                ref mut event,
+                updates,
+            } => {
+                *self = Cell::Full(box FullCell::Recomputing {
+                    event: event.take(),
+                    updates,
+                    dependent_tasks: AutoSet::from([reader]),
+                });
+            }
+            Cell::Full(box FullCell::Recomputing {
+                dependent_tasks, ..
+            })
+            | Cell::InitialValue {
+                dependent_tasks, ..
+            }
+            | Cell::Full(box FullCell::UpdatedValue {
+                dependent_tasks, ..
+            }) => {
+                dependent_tasks.insert(reader);
+            }
+        }
     }
 
     pub fn assign(&mut self, content: CellContent, turbo_tasks: &dyn TurboTasksBackendApi) {
-        self.content = content;
-        self.updates += 1;
-        // notify
-        if !self.dependent_tasks.is_empty() {
-            turbo_tasks.schedule_notify_tasks_set(&self.dependent_tasks);
+        match self {
+            Cell::Empty => {
+                *self = Cell::InitialValue {
+                    content,
+                    dependent_tasks: AutoSet::new(),
+                };
+            }
+            &mut Cell::Recomputing {
+                ref mut event,
+                updates,
+            } => {
+                event.notify(usize::MAX);
+                if updates == 1 {
+                    *self = Cell::InitialValue {
+                        content,
+                        dependent_tasks: AutoSet::new(),
+                    };
+                } else {
+                    *self = Cell::Full(box FullCell::UpdatedValue {
+                        content,
+                        dependent_tasks: AutoSet::new(),
+                        updates,
+                    });
+                }
```

Comment:
I can't find a path that would have a `Cell::Recomputing` with an `updates` value greater than 1. The only thing that seems to increment updates is a `FullCell::UpdatedValue`, but I don't think one of those can convert into anything else.